managed/services/onboarding/tips.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package onboarding provides functionality for user onboarding features.
+package onboarding
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm/api/onboardingpb"
+	"github.com/percona/pmm/managed/models"
+)
+
+const (
+	InstallPMMServerTipID    = 1
+	InstallPMMClientTipID    = 2
+	ConnectServiceToPMMTipID = 3
+)
+
+type TipsService struct {
+	db               *reform.DB
+	inventoryService inventoryService
+
+	systemTipIDs map[int32]struct{}
+
+	onboardingpb.UnimplementedTipServiceServer
+}
+
+var _ onboardingpb.TipServiceServer = (*TipsService)(nil)
+
+func NewTipService(db *reform.DB, inventoryService inventoryService) *TipsService {
+	return &TipsService{
+		db:               db,
+		inventoryService: inventoryService,
+		systemTipIDs: map[int32]struct{}{
+			InstallPMMServerTipID:    {},
+			InstallPMMClientTipID:    {},
+			ConnectServiceToPMMTipID: {},
+		},
+	}
+}
+
+func (t *TipsService) GetTipStatus(ctx context.Context, tipRequest *onboardingpb.GetTipfRequest) (*onboardingpb.GetTipResponse, error) {
+	switch tipRequest.TipType {
+	case onboardingpb.TipType_SYSTEM:
+		tip, err := t.retrieveSystemTip(tipRequest.TipId)
+		if err != nil {
+			return nil, err
+		}
+		return &onboardingpb.GetTipResponse{
+			TipId:       tip.ID,
+			IsCompleted: tip.IsCompleted,
+		}, nil
+	case onboardingpb.TipType_USER:
+		return t.retrieveOrCreateUserTip(tipRequest)
+	default:
+		return nil, errors.New("Tip type is not correct")
+	}
+}
+
+func (t *TipsService) retrieveOrCreateUserTip(tipRequest *onboardingpb.GetTipfRequest) (*onboardingpb.GetTipResponse, error) {
+	tip, err := t.retrieveUserTip(tipRequest.TipId, tipRequest.UserId)
+	if err != nil {
+		if err == reform.ErrNoRows {
+			tip, err = t.createUserTip(tipRequest.TipId, tipRequest.UserId)
```

Comment:
if I add constraint that user_id and user_tip_id should be unique, then this case won't be relevant, right? https://github.com/percona/pmm/blob/1e7805b988ac8a4db31b1769ee64ecb040ebc834/managed/models/database.go#L853

like this

```
CREATE TABLE user_tips (
		    id SERIAL PRIMARY KEY,
			user_id PRIMARY KEY INTEGER NOT NULL,
			user_tip_id INTEGER NOT NULL,
			is_completed BOOLEAN NOT NULL,
	
			created_at TIMESTAMP NOT NULL,
			updated_at TIMESTAMP NOT NULL
		   UNIQUE (user_id, user_tip_id)
		);
```

because in case if during request we would try to create user tip with existing ids then database should handle attempt of creation tip with the same id.

Reply:
added this constraint and wrapped this code into db transaction